### PR TITLE
synthetic-network for nightly-pipeline

### PR DESCRIPTION
**Pull request by fixxxedpoint**
# Description

- Rust API for the synthetic-network
- two e2e tests invoking synthetic-network (nightly pipeline)

## Type of change

tests/tooling

### CODE REVIEW

File: e2e-tests/src/config.rs [hunk 1]
```diff
@@ -87,13 +89,29 @@ impl Config {
             .unwrap()
     }
 
+    pub fn validator_names(&self) -> Vec<String> {
```
Comment (maciejzelaszczyk) [inline e2e-tests/src/config.rs:92]: Would parametrize this based on `validator_count`.

File: e2e-tests/src/synthetic_network.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+pub type Milliseconds = u64;
+
+fn create_client(node_name: &str) -> SyntheticNetworkClient {
+    let synthetic_network_url = format!("http://{}:80/qos", node_name);
+    info!("creating an http client for url {}", synthetic_network_url);
+    SyntheticNetworkClient::new(synthetic_network_url)
+}
+
+pub async fn set_out_latency(milliseconds: Milliseconds, node_name: &str) {
```
Comment (maciejzelaszczyk) [inline e2e-tests/src/synthetic_network.rs:12]: Would either rename this to sth like `create_client_with_latency` or provide the latency as a parameter to the constructor of `SyntheticNetworkClient`.

File: e2e-tests/src/test/high_latency.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            "setting out-latency of node {} to {}",
+            validator, config.test_case_params.out_latency
+        );
+        set_out_latency(config.test_case_params.out_latency, &validator).await;
+    }
+
+    let mut max_session = 0;
+    for connection in &connections {
+        let node_session = connection.get_session(None).await;
+        max_session = max(max_session, node_session);
```
Comment (maciejzelaszczyk) [inline e2e-tests/src/test/high_latency.rs:35]: This suggests that the current session can be different for different nodes. Why does this happen?

File: e2e-tests/src/test/high_latency.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+use aleph_client::{
+    pallets::session::SessionApi,
+    waiting::{AlephWaiting, BlockStatus},
+};
+use log::info;
+
+use crate::{config::setup_test, synthetic_network::set_out_latency};
+
+#[tokio::test]
+pub async fn high_out_latency() -> anyhow::Result<()> {
```
Comment (maciejzelaszczyk) [inline e2e-tests/src/test/high_latency.rs:12]: I think a doc description of what this test does and why this is relevant would be beneficial here.

File: e2e-tests/src/test/high_latency.rs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    let config = setup_test();
+
+    let connections = config.create_signed_connections().await;
+    info!("waiting for at least session 3");
+    for connection in &connections {
+        if connection.get_session(None).await < 3 {
```
Comment (maciejzelaszczyk) [inline e2e-tests/src/test/high_latency.rs:18]: Why is this check necessary? (Probably relates to my question about `max_session`)

File: e2e-tests/src/test/high_latency.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    for connection in connections {
+        connection
+            .wait_for_session(max_session + 2, BlockStatus::Finalized)
+            .await;
+    }
+    Ok(())
+}
+
+#[tokio::test]
+pub async fn no_quorum_without_high_out_latency() -> anyhow::Result<()> {
```
Comment (maciejzelaszczyk) [inline e2e-tests/src/test/high_latency.rs:48]: Test doc description.

File: scripts/synthetic-network/synthetic-link/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+#[derive(Serialize_repr, Deserialize_repr, Clone)]
+#[repr(u8)]
+pub enum Protocol {
+    Icmp = 1,
+    Tcp = 6,
+    Udp = 17,
+    All = 0,
+}
+
+/// Simple wrapper for the `RangeInclusive<u16` type.
```
Comment (maciejzelaszczyk) [inline scripts/synthetic-network/synthetic-link/src/lib.rs:166]: Lowly nit: `<u16>`.

**Review by maciejzelaszczyk — CHANGES_REQUESTED** [review]
Very nice indeed, couple of comments/questions.

File: e2e-tests/src/test/high_latency.rs (hunk 3 above)
Comment (fixxxedpoint) [reply e2e-tests/src/test/high_latency.rs:35]: Some can be on `+1` already?

File: e2e-tests/src/test/high_latency.rs (hunk 5 above)
Comment (fixxxedpoint) [reply e2e-tests/src/test/high_latency.rs:18]: it's not an issue - I can remove it, I needed it to speed local tests.

File: e2e-tests/src/test/high_latency.rs (hunk 3 above)
Comment (fixxxedpoint) [reply e2e-tests/src/test/high_latency.rs:35]: Anyway, one node can be even on arbitrary low round. Actually, that suggest that this test can be made less flaky/different - let them reach a session -> disconnect them so they stop proceeding -> set latency -> reconnect.

File: e2e-tests/src/synthetic_network.rs (hunk 2 above)
Comment (fixxxedpoint) [reply e2e-tests/src/synthetic_network.rs:12]: Why? Sounds more confusing for me. I am not creating a client with latency, I am setting latency of the network. This is just a `convenience-method` in spirit of our other methods like this. I can make `SyntheticNetworkClient` public here directly, but I see no point for doing that now.

File: shell.nix [hunk 8]
```diff
@@ -15,4 +15,5 @@ nixpkgs.mkShell.override { stdenv = env; }
   {
     inherit nativeBuildInputs;
     inherit (project) buildInputs shellHook;
+    RUST_SRC_PATH = "${versions.rustToolchain.rust-src}/lib/rustlib/src/rust/library/";
```
Comment (fixxxedpoint) [inline shell.nix:18]: I will a comment with explanation here.

**Review by maciejnems — APPROVED** [review]
Appart from what @maciejzelaszczyk said about docs it would be nice to add to README some description on how to run synthetic network e2e tests.

File: e2e-tests/src/synthetic_network.rs (hunk 2 above)
Comment (maciejzelaszczyk) [reply e2e-tests/src/synthetic_network.rs:12]: Maybe hand the `client` down as a dependency then? My main gripe here is that you are creating one from scratch in a `set_` function.

Comment (maciejzelaszczyk) [reply e2e-tests/src/synthetic_network.rs:12]: Not something blocking, though. If you feel strongly about this, resolve the comment.

File: e2e-tests/src/config.rs (hunk 1 above)
Comment (fixxxedpoint) [reply e2e-tests/src/config.rs]: done

File: e2e-tests/src/test/high_latency.rs (hunk 4 above)
Comment (fixxxedpoint) [reply e2e-tests/src/test/high_latency.rs:17]: added

File: e2e-tests/src/test/high_latency.rs (hunk 6 above)
Comment (fixxxedpoint) [reply e2e-tests/src/test/high_latency.rs:57]: added

File: scripts/synthetic-network/synthetic-link/src/lib.rs (hunk 7 above)
Comment (fixxxedpoint) [reply scripts/synthetic-network/synthetic-link/src/lib.rs]: fixed

**Review by maciejzelaszczyk — APPROVED** [review]

**Pull request merged**